File: python/binary_search_recursive.py


def binary_search(l: list, low: int, high: int, x: int):
    if high >= low:
        m = (high+low) // 2
        if l[m] == x:
            return m
        elif l[m] > x:
            binary_search(l, low, m-1, x)
        else:
            binary_search(l, m+1, high, x)
    else:
        return -1


def main():
    l = [1, 3, 5, 11, 104, 200, 1000]
    print(binary_search(l, 0, len(l)-1, 11))

main()

File: cpp/BinarySearch/binary_search_recursive.cpp
#include<iostream>
#include<vector>
using namespace std;


int binarySearch(vector<int> vec, int low, int high, int x) {
	int m = (low + high) / 2;
	if(x<vec[m]) {
		return binarySearch(vec, low, m-1, x);
	} 
	if(x>vec[m]) {
		return binarySearch(vec, m+1, high, x);
	} 
	if(x==vec[m]) {
		return m;
	}
	if(high==low) {
		return -1;
	}
}


int main() {
	vector<int> vec = {1, 3, 5, 10, 20, 100};
	cout << binarySearch(vec, 0, 5, 3) << "\n";
}


